### PR TITLE
Add the slack platform support

### PR DESCRIPTION
**Pull request by ash0521**
I hope this time will work,,,RUN

### CODE REVIEW

**Comment by BinaryHB0916** [comment]
Thanks Ash, community will check this by next Monday.

**Comment by ash0521** [comment]
Could you please build now?

**Comment by ash0521** [comment]
😪

**Comment by ash0521** [comment]
Please take a closer look and say something

**Comment by nykma** [comment]
I'll take a close look after CI is passed.
I suggest you to build and test it locally before committing

- Run `gofmt -l .` until no output
- Run `make build` until no error

If you have no clue what I'm saying, learn how to programming using `go` first.

**Comment by ash0521** [comment]
Sorry,I forget gofmt, please stay online

**Comment by ash0521** [comment]
Clear

**Comment by ash0521** [comment]
Don't run,I will all checked then text you,when totally resolved

**Comment by ash0521** [comment]
Man it has been a long time, Please finish

**Comment by nykma** [comment]
Good, but we still need a test to represent the whole workflow (at least validation workflow). Especially, I want to know the format of field `Identity` and `ProofLocation` (and the reason). They will be all self-explained in unit case so no need to reply to it here.

No need to put Slack API key in the test code, just code and run locally on your machine. Commit it when you're done.

> Take `validator/twitter/twitter_test.go` or `validator/minds/minds.go` as samples.

**Comment by ash0521** [comment]
Build

**Comment by ash0521** [comment]
sorry, i'm working.

**Comment by ash0521** [comment]
Please finish quickly

**Comment by ash0521** [comment]
Request Could you please finish the test today?

**Comment by ash0521** [comment]
Could you please finish my all job here by today

**Comment by ash0521** [comment]
Please finish my work first🙏🙏

File: validator/slack/slack.go
Comment (nykma) [inline validator/slack/slack.go:96]: What does `slack.ProofLocation` look like? I can't find any sample in your test case.

**Comment by ash0521** [comment]
Man, I can't be rest until a job is done

File: validator/slack/slack.go
Comment (nykma) [inline validator/slack/slack.go:112]: Not deal with pagination? What if target message is not found in the first page?

File: validator/slack/slack.go
Comment (nykma) [inline validator/slack/slack.go:122]: potential bug here: https://github.com/golang/go/issues/20733

File: validator/slack/slack_test.go
Comment (nykma) [inline validator/slack/slack_test.go:22]: Why mock it? This is the core function I want to test.

**Review by nykma — CHANGES_REQUESTED** [review]
Do you really understand what test case is for?

File: validator/slack/slack.go
Comment (ash0521) [reply validator/slack/slack.go:96]: Proof location will be url or channelID

Comment (ash0521) [reply validator/slack/slack.go:96]: Have you ever checked slack for prooflocation?

**Comment by ash0521** [comment]
Honestly, I lost my computer monitor.Suddenly has off while working.It will be a little late.

**Comment by ash0521** [comment]
ProofLocation will be https://slack.com/url/C12345678/12345678912345

**Comment by ash0521** [comment]
Done

**Comment by ash0521** [comment]
What do you know better about slack? Please take a closer look in my code and tell me details

File: validator/slack/slack_test.go
Comment (nykma) [inline validator/slack/slack_test.go:36]: What is this?

File: validator/slack/slack_test.go
Comment (nykma) [inline validator/slack/slack_test.go:37]: Where does this link come from? I can't open it in a browser.

File: validator/slack/slack_test.go
Comment (nykma) [inline validator/slack/slack_test.go:45]: Why you define `generate()` a second time?

**Review by nykma — CHANGES_REQUESTED** [review]
Some further thoughts about `slack` platform:

1. IMO `slack` has different account system on different `workspace`s, e.g., a same email can be registered in different `workspace`s and finally becomes different `account`s. Seems like you're not representing `workspace` in `Identity` field.
   > Or, do we really need to support `slack`? Sorry to raise this question up so late, but I'm not pretty sure about this. 
2. I still don't know what shape `ProofLocation` should be, Seems like the one in your test case is not legit.
3. You're terribly lacking of programming skills. We're not a playground / skill test platform and has no obligation to help you improve. Learn more by yourself, plan the proof validation procedure more clearly, then commit.

**Comment by ash0521** [comment]
Ah,This my public channel link,While a engineer added telegram the ProofLocation was added in the next.id docs

**Comment by ash0521** [comment]
Tell me a platform that you would easily accept

**Comment by ash0521** [comment]
You say ProofLocation,If we can't find ProofLocation, What is my efforts value!!!

**Comment by nykma** [comment]
Let's discuss these questions one by one.

1. In your design, how should users provide their `ProofLocation`?

**Comment by ash0521** [comment]
Honestly,I didn't pay to the attention on test case,Sorry for that,I follow telegram test code,you provide ProofLocation and he uses in their test code,I just remembered you told me before,
IMG_20230221_114329

**Comment by nykma** [comment]
> I follow telegram test code.

`platform`s differ widely, don't just copy / paste other `platform`'s impl / test code. You can take them as a reference, but duplicate them must not work.

So you still don't understand the whole workflow of how ProofService works. Learn it in [doc](https://docs.next.id/proof-service/ps-flow), and try to use it ([MaskBook](https://mask.io/) browser plugin's identity binding function) first. Until then, let's discuss about impl details.

**Comment by ash0521** [comment]
Check this,https://myworkspace.slack.com/archives/C1234567890/p123456789012345.

**Comment by nykma** [comment]
Also, don't post private chat into this public discussion.

**Comment by nykma** [comment]
> Check this,https://myworkspace.slack.com/archives/C1234567890/p123456789012345.

`C1234567890`? Is this real? Did you copy it from your `workspace`?

**Comment by ash0521** [comment]
No,This is example.

**Comment by nykma** [comment]
What I need is a **real** `ProofLocation` in the test. You should run this test locally on your computer, and submit it when it's OK. 

It will be better if you add some instruction of "how to get ProofLocation" in `README.md`.

**Comment by ash0521** [comment]
My public channel id,C04Q3P6H7TK

**Comment by nykma** [comment]
> My public channel id,C04Q3P6H7TK

Write it in test then.

**Comment by ash0521** [comment]
And you say do we need to add slack,Slack is better than discord,Big company uses slack,Most slack users are company employees.

**Comment by nykma** [comment]
> And you say do we need to add slack,Slack is better than discord,Big company uses slack,Most slack users are company employees.

I thought a little bit after that, I'm now realized that it worth to be supported, for on-premise scenario.

**Comment by ash0521** [comment]
Should we add page limit to our code?

**Comment by ash0521** [comment]
Please will you help me with the requirements of testing code?
IMG_20230223_215451

**Comment by ash0521** [comment]
I got ProofLocation shape, ProofLocation will be proof message link.

File: validator/slack/slack.go
Comment (nykma) [inline validator/slack/slack.go:110]: make it as a `const`

**Comment by ash0521** [comment]
Okay, Please help me in testing code
Check the minds testing reference
Can you please explain me the requirements of the testing code 🙏🙏
Can I get proof payload before code merge?
Also I can't generate signature in go
But ProofLocation and AltID is confirmed, ProofLocation will be proof message link and AltID will be UserID.

**Comment by ash0521** [comment]
Please tell🙏🙏

**Comment by ash0521** [comment]
Please would you cooperate to end it?

**Comment by nykma** [comment]
**Don't flood message in PR**. Github is not IM, you should NOT expect immediate response from others.

> Can you please explain me the requirements of the testing code

I think I already helped enough. Here is the final version:

1. Generate a keypair by your own, note down pubkey (you will use it in tests like [here](https://github.com/NextDotID/proof_server/blob/develop/validator/github/github_test.go#L15) )
2. Generate a sign payload (using your own `GenerateSignPayload()`) and a post payload (using your own `GeneratePostPayload()`) in your own `playground` mini program described below. Note down all necessary info to help you rebuild `validator.Base` in test case [like this](https://github.com/NextDotID/proof_server/blob/develop/validator/github/github_test.go#L22).
3. Sign the sign payload using private key generated in `1` , and place Base64 sig into post payload generated in `2`
4. Publish / send / post the post payload in real world (in your case, send it in a slack channel), like the common user will use ProofService finally. Copy the proof location (in your case, maybe the message share link. If message link doesn't have the info you need to `Validate()`, you need to find another way), and use it in your test [like this](https://github.com/NextDotID/proof_server/blob/develop/validator/github/github_test.go#L28).
5. Setup your slack API key locally, and run your test. If test went green, submit your code, not your API key.

> Can I get proof payload before code merge?

Write a small `cmd/playground`-ish by yourself, and generate one using your own `GeneratePostPayload()` func. Should be not that difficult.
BTW, Don't submit your `playground`, write and use it locally for your own.

> Also I can't generate signature in go

Impossible you can't generate sigs. There are many helper functions in `util/` and `util/crypto/` for you to use. 
Again, `cmd/playground` mini program just said before.

>  ProofLocation will be proof message link and AltID will be UserID.

`ProofLocation`, correct. `AltID`, wrong. See discussion in PR #45 for what `AltID` stands for.

**Comment by ash0521** [comment]
Check my Base64 sign:hvLhptjtazyIqj1FCyb/e6Chs/g8tD6gIg/YqyXSsQtQgit+WMmOakysKzpoPd+LpzlIhkGqVMH+OIsuvfdliQ==

**Comment by nykma** [comment]
I'M NOT A TEACHER CORRECTING YOUR HOMEWORK. REPRESENT YOUR RESULTS IN YOUR CODE.

**Comment by ash0521** [comment]
🥲

**Comment by ash0521** [comment]
Please take a look now

**Review by nykma — COMMENTED** [review]
Current code is OK to merge after CI turns green.

**Review by nykma — APPROVED** [review]

**Pull request merged**
